readthedocs/builds/storage.py

```diff
@@ -153,6 +155,19 @@ def sync_directory(self, source, destination):
                 log.debug('Deleting file from media storage.', filepath=filepath)
                 self.delete(filepath)
 
+    @cached_property
+    def _rclone(self):
+        raise NotImplementedError
+
+    def rclone_sync(self, source, destination):
+        """Sync a directory recursively to storage using rclone sync."""
+        if destination in ("", "/"):
+            raise SuspiciousFileOperation("Syncing all storage cannot be right")
```

Comment:
What are _valid_ destinations? Can't we check for them specifically?

Reply:
They can be anything, but syncing on the root of the storage shouldn't be allowed, we do this check in all sync operations.

Reply:
I was thinking about something like "allow a project to upload their files _only_ under specific paths". For example, `test-builds` will be able to sync directories _only_ at `{html,json,htmlzip,epub, pdf}/test-builds/<version>/`. Would that make sense for you?

Reply:
We could check for the first top-level dir, but if we are uploading something to a different top-level path, that won't affect anything. If we check for project and version, we then need to pass the project and version, and that feels like a check that should be done outside the storage class. With Eric we were discussing the possibility of using temporal AWS tokens for each user upload, haven't looked into it much, but looks like we could restrict the path with that.

Reply:
> With Eric we were discussing the possibility of using temporal AWS tokens for each user upload, haven't looked into it much, but looks like we could restrict the path with that.

👍🏼 --this is a good research to do at some point, yeah

Reply:
I will be good if you create an issue in -ops and add some context in the description, so we can come back to this research later.